runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
+  if (iree_status_is_ok(status)) {
+    // The current value stored in the semaphore is greater than the current
+    // value, so we can return.
+    if (semaphore->current_visible_value >= value) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_event_semaphore_advance(base_semaphore);
+    } else if (iree_timeout_is_infinite(timeout)) {
+      // This is the fast-path. Since we have an infinite timeout, we can
+      // wait directly on the hip event.
+
+      // The current value is not enough, but we have at least submitted
+      // the work that will increment the semaphore to the value we need.
+      // Use iree_hal_hip_util_tree_lower_bound to find the first element in the
+      // tree that would signal our semaphore to at least the given value.
+      iree_hal_hip_util_tree_node_t* node = iree_hal_hip_util_tree_lower_bound(
+          &semaphore->event_queue.tree, value);
+      IREE_ASSERT(
+          node,
+          "We really should either have an event in the queue that will satisfy"
+          "this semaphore, (we checked max_value_to_be_signaled above), or we"
+          "should already have signaled (current_visible_value above)");
+      iree_hal_hip_semaphore_queue_item_t* item =
+          (iree_hal_hip_semaphore_queue_item_t*)
+              iree_hal_hip_util_tree_node_get_value(node);
+
+      iree_hal_hip_event_t* event = item->event;
+
+      // Retain the event, as the event may be removed from the tree
+      // while we sleep on the event.
+      iree_hal_hip_event_retain(event);
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_event_semaphore_advance(base_semaphore);
+      status = IREE_HIP_CALL_TO_STATUS(
+          semaphore->symbols,
+          hipEventSynchronize(iree_hal_hip_event_handle(event)));
+      iree_hal_hip_event_release(event);
+    } else {
+      // If we have a non-infinite timeout, this is the slow-path.
+      // because we will end up having to wait for either the
+      // cleanup thread, or someone else to advance the
+      // semaphore.
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      iree_hal_hip_cpu_event_t* cpu_event = NULL;
+      status = iree_hal_hip_semaphore_get_cpu_event(base_semaphore, value,
+                                                    &cpu_event);
+      if (iree_status_is_ok(status)) {
+        // If there is no cpu event the semaphore has hit the value already.
+        if (cpu_event) {
+          status = iree_wait_one(&cpu_event->event, deadline_ns);
+          iree_hal_resource_release(&cpu_event->resource);
+        }
+      }
+    }
+  }
+
+  // If we are ok status, we expect that the mutex was unlocked.
+  // If there was an error the mutex is still locked.
+  if (iree_status_is_ok(status)) {
+    iree_slim_mutex_lock(&semaphore->mutex);
+    if (semaphore->current_visible_value >= IREE_HAL_SEMAPHORE_FAILURE_VALUE) {
+      status = iree_status_join(
```

Comment:
status is ok so you aren't joining anything - this should just be a status = iree_make_status.